src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -133,7 +133,11 @@ object ExpandWhens extends Pass {
                     case (WInvalid(), WInvalid()) => WInvalid()
                     case (WInvalid(), fv) => ValidIf(NOT(s.pred), fv, tpe(fv))
                     case (tv, WInvalid()) => ValidIf(s.pred, tv, tpe(tv))
-                    case (tv, fv) => Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                    case (tv, fv) => {
+                      // parameterization might result in the same tv, fv -- get rid of unnecessary mux
+                      if (tv != fv) Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                      else tv
```

Comment:
If basic constant propagation shouldn't be done here, then there definitely needs to be some utility mechanism to traverse the graph and figure out if 2 nodes have the same origin. I personally dislike the notion of always having to wait until the end to do constant propagation, because that makes intermediate FIRRTL code harder to decipher and potentially harder to do anything with. (If we want some conditional optimization). 

Also, by using the above + a ConstProp pass leads to:

```
assign mem_T_5_mask = 1'h1;

always @(posedge clk) begin
    GEN_1 <= mem_T_3_addr;
    GEN_3 <= mem_T_3_en;
    if(mem_T_5_en & mem_T_5_mask) begin
      mem[mem_T_5_addr] <= mem_T_5_data;
    end
  end
```

Which is considerably less redundant (given the fact that no mask is needed) than:

```
assign mem_T_5_mask = T_4;
assign mem_T_5_en = T_4;
assign T_4 = io_en & io_wen;

always @(posedge clk) begin
    GEN_1 <= mem_T_3_addr;
    GEN_3 <= mem_T_3_en;
    if(mem_T_5_en & mem_T_5_mask) begin
      mem[mem_T_5_addr] <= mem_T_5_data;
    end
  end
```

Which current FIRRTL generates. -- It seems like the ConstProp or other logic redunction pass didn't really do its job. While you could say that this optimization should be left for the synthesis tools, Verilog readibility + code line reduction is obviously still quite valuable to the hardware designer, so wherever possible, my opinion is you should try to sneak in optimizations that make sense. 

It's easier to catch areas for optimization tailored to specific points in the compilation process than it is to to create one big catch-all at the end...